File: 2020/23_crab_cups.rb
EXAMPLE = "389125467"
REAL = "418976235"
PT2 = true

input = REAL

class Array
    def link_each(starting=1)
        return to_enum(:link_each, starting) unless block_given?
        cur = starting
        loop do
            cur = self[cur]
            break if cur == starting
            yield cur
        end
    end
end
cups = Array.new(PT2 ? 1_000_001 : input.size+1, nil)

init = input.split(//).map(&:to_i)
cups[init[-1]] = init[0]
1.upto(init.size-1) { |i| cups[init[i-1]] = init[i] }

if PT2
    cups[init[-1]] = init.size+1
    (init.size+1).upto(1_000_000) { |i| cups[i] = i+1 }
    cups[1_000_000] = init[0]
end
cur = init[0]
min = 1
max = cups.size - 1

cycles = PT2 ? 10_000_000 : 100
cycles.times do |i|
    # puts "\n-- move #{i+1} --"

    # puts "cups: #{cups.link_each(cur).to_a.join(' ')}"
    pick = [cups[cur], cups[cups[cur]], cups[cups[cups[cur]]]]
    # pick = cups.link_each(cur).take(3)
    # puts "pick up: #{pick.join(", ")}"

    dest = cur - 1
    dest = max if dest < min
    while pick.include?(dest)
        dest -= 1
        dest = max if dest < min
    end
    # puts "destination: #{dest}"
    # cups.insert(desti+1, *pick)
    # cur = cups[(cups.index(cur) + 1) % cups.size]
    prev = cups[dest]
    cups[dest] = pick[0]
    cups[cur] = cups[pick[-1]]
    cups[pick[-1]] = prev
    cur = cups[cur]
end

if PT2
    a, b = cups.link_each.take(2)
    puts "#{a} * #{b} = #{a*b}"
else
    puts cups.link_each.to_a.join
end

File: 2020/23_crab_cups.go
package main

import "fmt"

// const input = "389125467" // example
const input = "418976235" // real
const pt2 = true

func main() {
	var moves, cuplen int
	if pt2 {
		moves = 10_000_000
		cuplen = 1_000_001
	} else {
		moves = 100
		cuplen = len(input) + 1
	}

	// parse input
	init := make([]int, len(input))
	for idx, c := range input {
		init[idx] = int(c - '0')
	}

	// build cups array
	cups := make([]int, cuplen)
	cups[init[len(init)-1]] = init[0]
	for idx := 1; idx < len(init); idx++ {
		cups[init[idx-1]] = init[idx]
	}
	if pt2 {
		cups[init[len(init)-1]] = len(init) + 1
		for idx := len(init) + 1; idx < 1_000_000; idx++ {
			cups[idx] = idx + 1
		}
		cups[1_000_000] = init[0]
	}

	cur := init[0]
	min := 1
	max := len(cups) - 1

	for move := 1; move <= moves; move++ {
		pick := [3]int{cups[cur], cups[cups[cur]], cups[cups[cups[cur]]]}
		picked := func(dest int) bool {
			return pick[0] == dest || pick[1] == dest || pick[2] == dest
		}

		var dest int
		for dest = cur; dest == cur || picked(dest); {
			dest--
			if dest < min {
				dest = max
			}
		}
		// fmt.Printf("\n-- move %d --\n", move)
		// fmt.Printf("cups: %v\n", cups)
		// fmt.Printf("pick up: %v\n", pick)
		// fmt.Printf("destination: %d\n", dest)
		prev := cups[dest]
		cups[dest] = pick[0]
		cups[cur] = cups[pick[2]]
		cups[pick[2]] = prev
		cur = cups[cur]
	}

	if pt2 {
		a := cups[1]
		b := cups[a]
		fmt.Printf("%d * %d = %d", a, b, a*b)
	} else {
		for cur = cups[1]; cur != 1; cur = cups[cur] {
			fmt.Print(cur)
		}
	}
	fmt.Println()

}
